### DISCUSSION TITLE
randomize order of dict methods items(), values(), keys()

### DISCUSSION THREAD

**Discussion by guettli** [pytest-dev/pytest#8934] [General] [answered]
I am looking at this code.

```
hashlib.md5(
            json.dumps(list(some_dict.values())).encode('utf-8')
        ).hexdigest()
```
Up to now the hash was stable. But this is pure luck.

Is there a way to make the dict methods items(), values() and keys() return the items in random order?

**Comment by nicoddemus** [comment, marked as answer]
@guettli not clear what you mean, and why this is related to pytest.

> But this is pure luck.

Since Python 3.7 dicts are ordered, so I'm not sure how this has changed for you.

**Reply by guettli** [reply]
Sorry, I missed that.

**Reply by nicoddemus** [reply]
No worries. 👍

**Reply by guettli** [reply]
Just for the records: the above example code has a flaw. The order of the items in the dictionary is an implementation detail, which can change. If you want a reproducible hash, then the input needs to be normalized before hashing. Otherwise the hash depends on the insertion order of the dictionary. It is unlikely that this is what you want.... But this not related to pytest at all.
